gix-worktree/src/diff.rs

```diff
@@ -0,0 +1,292 @@
+        self.stat_changed = self.stat_changed || index_entry.stat.mtime >= index.timestamp()
+    }
+
+    /// returns true if this entry has any changes
+    /// usually `detect_racy_stat` should be called first to avoid race condition
+    pub fn changed(&self) -> bool {
+        self.mode_change.is_some() || self.stat_changed || self.data_changed
+    }
+
+    /// Reads the worktree file from the disk and compares it to
+    /// the index entries oid to check if the actual data of the file is changed
+    /// and sets [`Entry::data_changed`] accordingly
+    pub fn compare_data(
+        &mut self,
+        worktree_path: &Path,
+        index_entry: &index::Entry,
+        buf: &mut Vec<u8>,
+        capabilities: &fs::Capabilities,
+    ) -> Result<(), read::Error> {
+        if self.mode_change.is_some() || !self.stat_changed || self.data_changed {
+            return Ok(());
+        }
+        let data = read_blob_to_buf_with_meta(
+            worktree_path,
+            index_entry.mode.contains(index::entry::Mode::SYMLINK),
+            buf,
+            capabilities,
+        )?;
+        let header = loose_header(gix_object::Kind::Blob, data.len());
+        let hash_changed = match index_entry.id {
+            ObjectId::Sha1(entry_hash) => {
+                let mut file_hash = hash::Sha1::default();
+                file_hash.update(&header);
+                file_hash.update(&data);
+                let file_hash = file_hash.digest();
+                entry_hash != file_hash
+            }
+        };
+        self.data_changed = hash_changed;
+        Ok(())
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Could not convert path to UTF8 {path}")]
+    IllformedUtf8 { path: BString },
+    #[error("The clock was off when reading file related metadata after updating a file on disk")]
+    Time(#[from] std::time::SystemTimeError),
+    #[error("IO error while writing blob or reading file metadata or changing filetype")]
+    Io(#[from] io::Error),
+}
+
+#[derive(Clone, Debug)]
+/// A change between the index and the worktree computed by [`compate_to_index`]
+pub struct Change<'a> {
+    /// The index entry that changed
+    pub index_entry: &'a index::Entry,
+    /// The on-disk worktree path corresponding to this entry
+    pub worktree_path: PathBuf,
+    /// How this index entry changed
+    pub kind: ChangeKind,
+    /// file metadata that can be reused (optimization)
+    pub fstat: Option<std::fs::Metadata>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+///
+pub enum ChangeKind {
+    /// An index entry has no corresponding file in the worktree
+    Removed,
+    /// Ar new files that has been marked with git add but has not yet been
+    /// checked in yet. No diff is computed for these files because whatever is
+    /// on disk at commit time will be used
+    Added,
```

Comment:
I have trouble understanding this variant in the context of `index-to-worktree` diff as it seems to imply a comparison with the commit's tree. In theory, `Added` should be 'it exists in the index' but is deleted in the worktree?

Reply:
New (otherwise untracked files) get added to the index with `git add` so in theory this file may or may or may not have changed compared to the index. However this is not usually what the user "case" about instead these files are simply shown as added (in git status, in git diff they are simply shown as all green).

This is detected by a special flag stored in the index.